sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1208,9 +1208,26 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
             reduceLeftOption(And).map(Filter(_, left)).getOrElse(left)
           val newRight = rightJoinConditions.
             reduceLeftOption(And).map(Filter(_, right)).getOrElse(right)
-          val newJoinCond = commonJoinCondition.reduceLeftOption(And)
-
-          Join(newLeft, newRight, joinType, newJoinCond)
+          val (newJoinConditions, others) =
+            commonJoinCondition.partition(canEvaluateWithinJoin)
+          val newJoinCond = newJoinConditions.reduceLeftOption(And)
+          // if condition expression is unevaluable, it will be removed from
+          // the new join conditions, if all conditions is unevaluable, we should
+          // change the join type to CrossJoin.
+          val newJoinType =
+            if (commonJoinCondition.nonEmpty && newJoinCond.isEmpty) {
+              logWarning(s"The whole commonJoinCondition:$commonJoinCondition of the join " +
+                s"plan:\n $j is unevaluable, it will be ignored and the join plan will be " +
```

Comment:
not sure that inlining the plan here makes this warning very readable...

Reply:
The log will be shown like this:
```
16:13:35.218 WARN org.apache.spark.sql.catalyst.optimizer.PushPredicateThroughJoin: The whole commonJoinCondition:List((dummyUDF(a#5, b#6) = dummyUDF(d#15, c#14))) of the join plan:
 Join Inner, (dummyUDF(a#5, b#6) = dummyUDF(d#15, c#14))
:- LocalRelation [a#5, b#6]
+- LocalRelation [c#14, d#15]
 is unevaluable, it will be ignored and the join plan will be turned to cross join.
```

Reply:
@xuanyuanking @mgaido91 In the above example, the UDF's refer to attributes from distinct legs of the join. Can we not plan this better than a cross join in this case ? I am wondering why we can't do -
```
  Join Inner, leftAlias1 = rightAlias1
      Project dummyUDF(a, b) as leftAlias1
         LocalRelation(a, b)
      Project dummyUDF(c, d) as rightAlias1
         LocalRelation(c, d)
```
Perhaps i am missing something ..

Reply:
@dilipbiswal I haven't checked the particular plan posted in that comment, for which I think you are right, we can handle as you suggested, but I was checking the case in the UT and in the description of this PR, ie. when the input for the Python UDF contains attributes from both sides. In that case I don't have a better suggestion.

Reply:
@mgaido91 Thanks. Marco, do you know if there are instances when we pick cross join implicitly ? It wouldn't perform very well, right ? Wondering if we should error out or pick a bad plan. I guess, i am not sure whats the right thing to do here.

Reply:
@dilipbiswal Thanks for your detailed check, I should write the case more typical, here the case we want to solve is UDF which accessing the attribute in both side, I'll change the case to `dummyPythonUDF(col("a"), col("c")) === dummyPythonUDF(col("d"), col("c"))` in next commit.

Reply:
```
yes, and if the plan is big, than this would become quite unreadable IMHO. I think it would be better to refactor the message and put the plan at the end.
```
@mgaido91 Thanks for your advise, will do the refactor in next commit.

Reply:
@dilipbiswal there are cases when "trivial conditions" are removed from a join so we make a inner join a cross one for instance. The performance would be awful, you're right. The point is that I am not sure that there is a better way to achieve this. I mean, since we have no clue what the UDF does, we need to compare all the rows from both sides, ie. we need to perform a cartesian product.

> Wondering if we should error out or pick a bad plan

This is, indeed, arguable. I think that letting the user choose is a good idea. If the users runs the query and gets an `AnalysisException` because he/she is trying to perform a cartesian product, he/she can decide: ok, I am doing a wrong thing, let's change it; or he/she can say, well, one of my 2 tables involved contains 10 rows, not a big deal, I want to perform it nonetheless, let's set `spark.sql.crossJoin.enabled=true` and run it.

> for join types other than inner and leftsemi, we still have the same issue, no ?

I think the current PR handles properly only the case with type inner (for the left semi) this PR returns an incorrect result IIUC. This needs to be fixed as well.